src/qs8-gemm/neon-mlal-lane.c.in

```diff
@@ -127,230 +153,294 @@ void xnn_${DATATYPE_SPEC}_gemm_minmax${REQUANTIZATION_SPEC}_ukernel_${MR}x${NR}_
+                    ${_}vacc${M}x${ABC[N+4:N+8]} = vmlal_lane_s16(vacc${M}x${ABC[N+4:N+8]}, vget_high_s16(vxb${ABC[N:N+8]}c5), vget_high_s16(vxa${M}), 1);
 
-                $if DATATYPE not in ["QC4_F16", "QC4_F32"]:
-                  if (k > 6 * sizeof(${XINT8_T})) {
+                $if DATATYPE not in ["QC4_F16", "QC4_F32", "QB4_F16", "QB4_F32"]:
+                  ${_}if (k > 6 * sizeof(${XINT8_T})) {
                     $for N in range(0, NR, 8):
-                      const ${XINT8X8_T} vb${ABC[N:N+8]}c6 = ${VLD1_X8}(w); w = (const ${XINT8_T}*) w + 8;
+                      ${_}const ${XINT8X8_T} vb${ABC[N:N+8]}c6 = ${VLD1_X8}(w); w = (const ${XINT8_T}*) w + 8;
                       $if DATATYPE == "QU8":
-                        const int16x8_t vxb${ABC[N:N+8]}c6 = vreinterpretq_s16_u16(vsubl_u8(vb${ABC[N:N+8]}c6, vb_zero_point));
+                        ${_}const int16x8_t vxb${ABC[N:N+8]}c6 = vreinterpretq_s16_u16(vsubl_u8(vb${ABC[N:N+8]}c6, vb_zero_point));
                       $else:
-                        const int16x8_t vxb${ABC[N:N+8]}c6 = vmovl_s8(vb${ABC[N:N+8]}c6);
+                        ${_}const int16x8_t vxb${ABC[N:N+8]}c6 = vmovl_s8(vb${ABC[N:N+8]}c6);
 
                     $for M in range(MR):
                       $for N in range(0, NR, 8):
-                        vacc${M}x${ABC[N:N+4]} = vmlal_lane_s16(vacc${M}x${ABC[N:N+4]}, vget_low_s16(vxb${ABC[N:N+8]}c6), vget_high_s16(vxa${M}), 2);
-                        vacc${M}x${ABC[N+4:N+8]} = vmlal_lane_s16(vacc${M}x${ABC[N+4:N+8]}, vget_high_s16(vxb${ABC[N:N+8]}c6), vget_high_s16(vxa${M}), 2);
-                  }
-              }
-            }
-          }
-        }
-      }
-    }
+                        ${_}vacc${M}x${ABC[N:N+4]} = vmlal_lane_s16(vacc${M}x${ABC[N:N+4]}, vget_low_s16(vxb${ABC[N:N+8]}c6), vget_high_s16(vxa${M}), 2);
+                        ${_}vacc${M}x${ABC[N+4:N+8]} = vmlal_lane_s16(vacc${M}x${ABC[N+4:N+8]}, vget_high_s16(vxb${ABC[N:N+8]}c6), vget_high_s16(vxa${M}), 2);
+                  ${_}}
+              ${_}}
+            ${_}}
+          ${_}}
+        ${_}}
+      ${_}}
+    ${_}}
+
+    $if BLOCKWISE:
+      $for N in range(0, NR, 4):
+        const float32x4_t vfilter_output_scale${ABC[N:N+4]} = vld1q_f32(w); w = (const float*) w + 4;
 
-    $if DATATYPE in ["QD8", "QC4_F16", "QC4_F32"]:
-      $for M in range(0, MR):
+      $for M in range(MR):
         $for N in range(0, NR, 4):
-          $if DATATYPE in ["QC4_F16", "QC4_F32"]:
-            float32x4_t vout${M}x${ABC[N:N+4]} = vcvtq_n_f32_s32(vacc${M}x${ABC[N:N+4]}, 4);
-          $else:
-            float32x4_t vout${M}x${ABC[N:N+4]} = vcvtq_f32_s32(vacc${M}x${ABC[N:N+4]});
+          const float32x4_t vf${M}x${ABC[N:N+4]} = vcvtq_f32_s32(vacc${M}x${ABC[N:N+4]});
 
-      $for M in range(0, MR, 2):
-        $if M + 1 == MR:
-          const float32x4_t vinput_scale${M} = vld1q_dup_f32(&quantization_params[${M}].inv_scale);
+        $for N in range(0, NR, 4):
```

Comment:
Please wrap vfmaq_f32 in XNN_ARCH_ARM64 like in

https://github.com/google/XNNPACK/blob/e4dc2e7f2eafb7af4d6d9b6ac5e03e0f6f139bea/src/qs8-igemm/neon-mlal-lane.c.in#L297

Reply:
Done. Good catch.